File: frontend/src/domain/Visual.ts
export interface VisualCardProps {
  slug: string
  title: string
  kind: string
  summaryMd: string
  attribution: string
  thumbnailSource: string | null
  assetPath: string | null
  needsSandbox: boolean
}

export class VisualCard {
  readonly slug: string
  readonly title: string
  readonly kind: string
  readonly summaryMd: string
  readonly attribution: string
  // Size-gated (see backend THUMBNAIL_MAX_BYTES) — null for `image` (whose
  // real content is `assetPath`, not `source`) or when the real source is
  // too big for a card-sized preview to be worth sending.
  readonly thumbnailSource: string | null
  readonly assetPath: string | null
  // Same discriminator VisualDetail uses — a card can't attempt a live
  // sandboxed preview for a code-bearing kind without knowing that's what
  // it is.
  readonly needsSandbox: boolean

  constructor(props: VisualCardProps) {
    this.slug = props.slug
    this.title = props.title
    this.kind = props.kind
    this.summaryMd = props.summaryMd
    this.attribution = props.attribution
    this.thumbnailSource = props.thumbnailSource
    this.assetPath = props.assetPath
    this.needsSandbox = props.needsSandbox
  }
}

export interface ResourceProps {
  id: number
  label: string
  url: string
  kind: string
}

export class Resource {
  readonly id: number
  readonly label: string
  readonly url: string
  readonly kind: string

  constructor(props: ResourceProps) {
    this.id = props.id
    this.label = props.label
    this.url = props.url
    this.kind = props.kind
  }
}

export interface VisualCategoryRef {
  slug: string
  name: string
  color: string
}

export interface VisualDetailProps {
  slug: string
  title: string
  kind: string
  source: string
  assetPath: string | null
  themeAffinity: string
  origin: string
  generator: string | null
  createdOn: string | null
  context: string
  course: string | null
  summaryMd: string
  notesMd: string
  needsSandbox: boolean
  attribution: string
  status: string
  category: VisualCategoryRef
  tags: string[]
  resources: Resource[]
}

// The full `/v/:slug` shape — everything VisualCardSchema sends plus the
// fields only the detail page needs (source, notes, resources, ...). A
// separate class from VisualCard rather than a superset/subclass: the two
// pages that use them (the preview grid, the detail page) genuinely want
// different, non-overlapping shapes, and a shared base would just be the
// handful of fields they happen to both have.
export class VisualDetail {
  readonly slug: string
  readonly title: string
  readonly kind: string
  readonly source: string
  readonly assetPath: string | null
  readonly themeAffinity: string
  readonly origin: string
  readonly generator: string | null
  readonly createdOn: string | null
  readonly context: string
  readonly course: string | null
  readonly summaryMd: string
  readonly notesMd: string
  readonly needsSandbox: boolean
  readonly attribution: string
  // "pending" vs "published" — what the submission actually did with the
  // request, not just whether a write key was typed into the form (a
  // *wrong* key still lands pending; only the server knows which happened).
  readonly status: string
  readonly category: VisualCategoryRef
  readonly tags: string[]
  readonly resources: Resource[]

  constructor(props: VisualDetailProps) {
    this.slug = props.slug
    this.title = props.title
    this.kind = props.kind
    this.source = props.source
    this.assetPath = props.assetPath
    this.themeAffinity = props.themeAffinity
    this.origin = props.origin
    this.generator = props.generator
    this.createdOn = props.createdOn
    this.context = props.context
    this.course = props.course
    this.summaryMd = props.summaryMd
    this.notesMd = props.notesMd
    this.needsSandbox = props.needsSandbox
    this.attribution = props.attribution
    this.status = props.status
    this.category = props.category
    this.tags = props.tags
    this.resources = props.resources
  }
}

File: backend/app/schemas/visual.py
from marshmallow import Schema, fields, validate

from app.models.enums import Context, Origin, Status, ThemeAffinity, VisualKind
from app.schemas.resource import ResourceCreateSchema, ResourceSchema

# Sends `source` on the card for any kind under this size, so the grid can
# attempt a real preview instead of a bare kind-name placeholder — a visual
# whose source is too big just falls back to that placeholder (or the
# frontend's ascii-art cover) on the card, same as before this existed for
# non-svg kinds. See docs/DECISIONS.md.
THUMBNAIL_MAX_BYTES = 20_000


def _values(enum_cls):
    return [member.value for member in enum_cls]


class CategoryRefSchema(Schema):
    slug = fields.Str(dump_only=True)
    name = fields.Str(dump_only=True)
    color = fields.Str(dump_only=True)


class VisualCardSchema(Schema):
    slug = fields.Str(dump_only=True)
    title = fields.Str(dump_only=True)
    kind = fields.Str(dump_only=True)
    theme_affinity = fields.Str(dump_only=True)
    summary_md = fields.Str(dump_only=True)
    created_on = fields.Date(dump_only=True, allow_none=True)
    needs_sandbox = fields.Bool(dump_only=True)
    attribution = fields.Str(dump_only=True)
    category = fields.Nested(CategoryRefSchema, dump_only=True)
    tags = fields.Method("get_tag_names", dump_only=True)
    # `image`'s actual content is a file, not `source` — cheap to send on
    # every card (it's just a path the browser requests, not the asset
    # bytes themselves), unlike `thumbnail_source` below which is
    # size-gated.
    asset_path = fields.Str(dump_only=True, allow_none=True)
    thumbnail_source = fields.Method("get_thumbnail_source", dump_only=True)

    def get_tag_names(self, visual):
        return [tag.name for tag in visual.tags]

    def get_thumbnail_source(self, visual):
        if visual.kind == VisualKind.IMAGE:
            return None
        if len(visual.source or "") < THUMBNAIL_MAX_BYTES:
            return visual.source
        return None


class VisualDetailSchema(VisualCardSchema):
    source = fields.Str(dump_only=True)
    origin = fields.Str(dump_only=True)
    generator = fields.Str(dump_only=True, allow_none=True)
    context = fields.Str(dump_only=True)
    course = fields.Str(dump_only=True, allow_none=True)
    notes_md = fields.Str(dump_only=True)
    status = fields.Str(dump_only=True)
    resources = fields.Nested(ResourceSchema, many=True, dump_only=True)
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)


class VisualCreateSchema(Schema):
    title = fields.Str(required=True, validate=validate.Length(min=1, max=160))
    kind = fields.Str(required=True, validate=validate.OneOf(_values(VisualKind)))
    source = fields.Str(load_default="")
    asset_path = fields.Str(load_default=None, allow_none=True)
    theme_affinity = fields.Str(
        load_default=ThemeAffinity.ADAPTIVE.value, validate=validate.OneOf(_values(ThemeAffinity))
    )
    origin = fields.Str(load_default=Origin.HUMAN.value, validate=validate.OneOf(_values(Origin)))
    generator = fields.Str(load_default=None, allow_none=True, validate=validate.Length(max=80))
    created_on = fields.Date(load_default=None, allow_none=True)
    context = fields.Str(load_default=Context.PERSONAL.value, validate=validate.OneOf(_values(Context)))
    course = fields.Str(load_default=None, allow_none=True, validate=validate.Length(max=60))
    summary_md = fields.Str(load_default="")
    notes_md = fields.Str(load_default="")
    category_slug = fields.Str(required=True)
    tags = fields.List(fields.Str(), load_default=list)
    resources = fields.List(fields.Nested(ResourceCreateSchema), load_default=list)


class VisualUpdateSchema(Schema):
    title = fields.Str(validate=validate.Length(min=1, max=160))
    kind = fields.Str(validate=validate.OneOf(_values(VisualKind)))
    source = fields.Str()
    asset_path = fields.Str(allow_none=True)
    theme_affinity = fields.Str(validate=validate.OneOf(_values(ThemeAffinity)))
    origin = fields.Str(validate=validate.OneOf(_values(Origin)))
    generator = fields.Str(allow_none=True, validate=validate.Length(max=80))
    created_on = fields.Date(allow_none=True)
    context = fields.Str(validate=validate.OneOf(_values(Context)))
    course = fields.Str(allow_none=True, validate=validate.Length(max=60))
    summary_md = fields.Str()
    notes_md = fields.Str()
    category_slug = fields.Str()
    tags = fields.List(fields.Str())


class VisualStatusSchema(Schema):
    status = fields.Str(required=True, validate=validate.OneOf(_values(Status)))
